Doc/c-api/lifecycle.rst

```diff
@@ -0,0 +1,236 @@
+       :c:member:`~PyTypeObject.tp_new` to allocate the memory for the new
+       object.
+     * :c:member:`~PyTypeObject.tp_init` initializes the newly created object.
+       :c:member:`!tp_init` can be called again to re-initialize an object, if
+       desired.
+
+   * After :c:member:`!tp_init` completes, the object is ready to use.
+   * After the last reference to an object is removed:
+
+     #. If an object is not marked as *finalized*, it might be finalized by
+        marking it as *finalized* and calling its
+        :c:member:`~PyTypeObject.tp_finalize` function.  :term:`CPython`
+        currently does not finalize an object when the last reference to it is
+        deleted, but this may change in the future.
+     #. If the object is marked as finalized,
+        :c:member:`~PyTypeObject.tp_clear` might be called to clear references
+        held by the object.  :term:`CPython` currently does not clear an object
+        in response to the deletion of the last reference, but this may change
+        in the future.
+     #. :c:member:`~PyTypeObject.tp_dealloc` is called to destroy the object.
+
+   * The :c:member:`~PyTypeObject.tp_finalize` function can optionally add a
+     reference to the object, *resurrecting* it and preventing its pending
+     destruction.  (Only :c:member:`!tp_finalize` can resurrect an object;
+     :c:member:`~PyTypeObject.tp_clear` and
+     :c:member:`~PyTypeObject.tp_dealloc` cannot.)  Resurrecting an object may
+     or may not cause the object's *finalized* mark to be removed.  Currently,
+     :term:`CPython` does not remove the *finalized* mark from a resurrected
+     object if the object's type supports garbage collection (i.e., the
+     :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set) but does remove the mark if the
+     object does not support garbage collection; either or both of these
+     behaviors may change in the future.
+   * :c:member:`~PyTypeObject.tp_dealloc` can optionally call
+     :c:member:`~PyTypeObject.tp_finalize` via
+     :c:func:`PyObject_CallFinalizerFromDealloc` if it wishes to reuse that
+     code to help with object destruction.  This is recommended because it
+     guarantees that :c:member:`!tp_finalize` is always called before
+     destruction.
+   * When :c:member:`~PyTypeObject.tp_dealloc` finishes object destruction, it
+     directly calls :c:member:`~PyTypeObject.tp_free` to deallocate the memory.
+   * If the object is a member of a :term:`cyclic isolate` and
+     :c:member:`~PyTypeObject.tp_clear` fails to break the reference cycle (or
+     that function is not called), the objects remain indefinitely
+     uncollectable (they "leak").  See :data:`gc.garbage`.
+
+   If the object is marked as supporting garbage collection (the
+   :c:macro:`Py_TPFLAGS_HAVE_GC` flag is set in
+   :c:member:`~PyTypeObject.tp_flags`), the following events are also possible:
+
+   * The garbage collector occasionally calls
```

Comment:
Nitpick: I think "periodically" is a better term here.
```suggestion
   * The garbage collector periodically calls
```

Reply:
I intentionally avoided "periodically" here because I don't want the reader to assume dependable regularity.  Generational GC means that sometimes an object is scanned and sometimes not, and users don't have much control over frequency or regularity.